Mathlib/Topology/MetricSpace/CoveringNumbers.lean

```diff
@@ -0,0 +1,211 @@
+
+## References
+
+* [R. Vershynin, *High-dimensional probability*][vershynin2018high]
+
+-/
+
+@[expose] public section
+
+open EMetric Set
+open scoped ENNReal NNReal
+
+namespace Metric
+
+variable {X : Type*} [PseudoEMetricSpace X] {A B C : Set X} {ε δ : ℝ≥0} {x : X}
+
+section Definitions
+
+/-- The extenal covering number of a set `A` in `X` for radius `ε` is the minimal cardinal of
+an `ε`cover by points in `X` (not necessarily in `A`). -/
+noncomputable
+def externalCoveringNumber (ε : ℝ≥0) (A : Set X) : ℕ∞ :=
+  ⨅ (C : Set X) (_ : IsCover ε A C), C.encard
+
+/-- The covering number(or internal covering number) of a set `A` for radius `ε` is
+the minimal cardinal of an `ε`cover contained in `A`. -/
+noncomputable
+def coveringNumber (ε : ℝ≥0) (A : Set X) : ℕ∞ :=
+  ⨅ (C : Set X) (_ : C ⊆ A) (_ : IsCover ε A C), C.encard
+
+/-- The packing number of a set `A` for radius `ε` is the maximal cardinal of an `ε`-separated set
+in `A`. -/
+noncomputable
+def packingNumber (ε : ℝ≥0) (A : Set X) : ℕ∞ :=
+  ⨆ (C : Set X) (_ : C ⊆ A) (_ : IsSeparated ε C), C.encard
+
+end Definitions
+
+@[simp]
+lemma externalCoveringNumber_empty (ε : ℝ≥0) : externalCoveringNumber ε (∅ : Set X) = 0 := by
+  simp [externalCoveringNumber]
+
+@[simp]
+lemma coveringNumber_empty (ε : ℝ≥0) : coveringNumber ε (∅ : Set X) = 0 := by simp [coveringNumber]
+
+@[simp]
+lemma externalCoveringNumber_eq_zero :
+    externalCoveringNumber ε A = 0 ↔ A = ∅ := by simp [externalCoveringNumber]
+
+@[simp]
+lemma coveringNumber_eq_zero : coveringNumber ε A = 0 ↔ A = ∅ := by simp [coveringNumber]
+
+lemma externalCoveringNumber_le_encard (hC : IsCover ε A C) :
+    externalCoveringNumber ε A ≤ C.encard := iInf₂_le C hC
+
+lemma coveringNumber_le_encard (h_subset : C ⊆ A) (hC : IsCover ε A C) :
+    coveringNumber ε A ≤ C.encard := (iInf₂_le C h_subset).trans (iInf_le _ hC)
+
+lemma externalCoveringNumber_anti (h : ε ≤ δ) :
+    externalCoveringNumber δ A ≤ externalCoveringNumber ε A := by
+  simp_rw [externalCoveringNumber]
+  gcongr
+  exact iInf_const_mono (fun h_cover ↦ h_cover.mono_radius h)
+
+lemma coveringNumber_anti (h : ε ≤ δ) : coveringNumber δ A ≤ coveringNumber ε A := by
+  simp_rw [coveringNumber]
+  gcongr
+  exact iInf_const_mono (fun h_cover ↦ h_cover.mono_radius h)
+
+lemma externalCoveringNumber_mono_set (h : A ⊆ B) :
+    externalCoveringNumber ε A ≤ externalCoveringNumber ε B := by
+  simp only [externalCoveringNumber, le_iInf_iff]
+  exact fun C hC ↦ iInf_le_of_le C <| iInf_le_of_le (hC.anti h) le_rfl
+
+lemma coveringNumber_eq_one_of_ediam_le (h_nonempty : A.Nonempty) (hA : EMetric.diam A ≤ ε) :
+    coveringNumber ε A = 1 := by
+  refine le_antisymm ?_ ?_
+  · have ⟨a, ha⟩ := h_nonempty
+    calc coveringNumber ε A
+      _ ≤ ({a} : Set X).encard :=
+        coveringNumber_le_encard (by simp [ha]) (.singleton_of_ediam_le hA ha)
+      _ ≤ 1 := by simp
+  · by_contra! h
+    rw [ENat.lt_one_iff_eq_zero] at h
+    refine h_nonempty.ne_empty (by simpa using h)
```

Comment:
Here you could use that lemma `A.Nonempty → 0 < coveringNumber ε A` I mentioned.